torch/distributed/device_mesh.py

```diff
@@ -785,10 +750,30 @@ def __getitem__(
                 # yet and need a follow-up,
                 # TODO: compiler + device_mesh slicing.
                 with torch._subclasses.fake_tensor.unset_fake_temporarily():
-                    submesh = _mesh_resources.create_sub_mesh(
-                        self, mesh_dim_names, slice_mesh_dims
+                    # When users sliced dim_names outside from current mesh, we will check whether
+                    # there is layout overlap. Eventually we will just directly throw error here because
+                    # we will deprecate the slicing of flattened dim_name from root mesh.
+                    if set(self._layouts) < set(layouts_sliced):
+                        sizes = tuple(x for l in layouts_sliced for x in l.sizes)
+                        strides = tuple(x for l in layouts_sliced for x in l.strides)
+                        base = torch.empty(sizes, dtype=torch.uint8)
+                        t = torch.as_strided(base, size=sizes, stride=strides)
+                        if torch._debug_has_internal_overlap(t):
+                            raise RuntimeError(
+                                f"slicing overlapping dim_names {mesh_dim_names} is not allowed"
+                            )
+                    res_mesh = DeviceMesh._from_layouts(
+                        self.device_type,
+                        tuple(layouts_sliced),
```

Comment:
i mentioned this in the following PR, but having a 'type' for this would be nice.

Reply:
This one is not needed now.